### DISCUSSION TITLE
Mesh Attribute Position: Vec<[f32;3]> vs Vec<f32>

### DISCUSSION THREAD

**Discussion by afonsolage** [bevyengine/bevy#2651] [Q&A] [answered]
So I'm playing with custom Mesh creation and I noticed something odd. First I created a plane using just a `Vec<f32>` with all floats needed and no error was shown but nothing is rendered neither.

```
    let v_f32 = vec![
        -0.5, -0.5, -0.5,
        0.5, -0.5, -0.5,
        0.5, 0.5, -0.5,
        -0.5, 0.5, -0.5,
    ];
    
    let indices = vec![0, 1, 2, 2, 3, 0];
    
    let mut mesh = Mesh::new(PrimitiveTopology::TriangleList);
    mesh.set_indices(Some(Indices::U32(indices)));
    mesh.set_attribute(Mesh::ATTRIBUTE_POSITION, v_f32);
    
    commands.spawn_bundle(MeshBundle {
        mesh: meshes.add(mesh),
        render_pipelines: RenderPipelines::from_pipelines(vec![RenderPipeline::new(
            pipeline_handle,
        )]),
        ..Default::default()
    });
```

But after reading the examples for a while I noticed that `Mesh::ATTRIBUTE_POSITION` expectes a `Vec<[f32;3]` so after just changing the `v_f32` to `v_arr` it worked fine:

```
    let v_arr = vec![
        [-0.5, -0.5, -0.5],
        [0.5, -0.5, -0.5],
        [0.5, 0.5, -0.5],
        [-0.5, 0.5, -0.5],
    ];
    
    let indices = vec![0, 1, 2, 2, 3, 0];
    
    let mut mesh = Mesh::new(PrimitiveTopology::TriangleList);
    mesh.set_indices(Some(Indices::U32(indices)));
    mesh.set_attribute(Mesh::ATTRIBUTE_POSITION, v_arr); //Vec<[f32;3]> now
    
    commands.spawn_bundle(MeshBundle {
        mesh: meshes.add(mesh),
        render_pipelines: RenderPipelines::from_pipelines(vec![RenderPipeline::new(
            pipeline_handle,
        )]),
        ..Default::default()
    });
```

My question is if `Mesh::ATTRIBUTE_POSITION` only works with `Vec<[f32;3]>` why it let me use `Vec<f32>` and no error/warning is shown? This is an intended behaviour?

I [placed an example on Gist](https://gist.github.com/afonsolage/5f4e848f8497a5bf702b686613d93f0a) if anyone interested.

**Comment by cart** [comment, marked as answer]
We need to support every possible attribute value in `mesh::set_attribute`. The "position" attribute might expect a `Vec<[f32;3]>`, but the "uv" attribute expects a `Vec<[f32;2]>`.  We accepted `Vec<f32>` because that is a valid attribute type (just not the right one for "position"). That being said, this _should_ probably be a runtime warning / error. The rendering issue is probably because the shader consuming this expects a `vec3` type, but the mesh is providing a `float` type. This is where the "mismatch" occurs and where the validation should happen. We should try to fix this in the upcoming renderer rework (slated to land in Bevy 0.6).
